posts/2024-05-05-check-cfg.md

```diff
@@ -0,0 +1,124 @@
+author: Urgau
+team: The Cargo Team <https://www.rust-lang.org/governance/teams/dev-tools#cargo>
+---
+
+The Cargo and Compiler team are delighted to announce that starting with Rust 1.80 (or nightly-2024-05-05) every _reachable_ `#[cfg]` will be **automatically checked** that they match the **expected config names and values**.
+
+This can help with verifying that the crate is correctly handling conditional compilation for different target platforms or features. It ensures that the cfg settings are consistent between what is intended and what is used, helping to catch potential bugs or errors early in the development process.
+
+This addresses a common pitfall for new and advanced users.
+
+This is another step to our commitment to provide user-focused tooling and we are eager and excited to finally see it fixed, after more than two years since the original [RFC 3013](https://github.com/rust-lang/rfcs/pull/3013)[^1].
+
+[^1]: The stabilized implementation and RFC 3013 diverge significantly, in particular there is only one form for `--check-cfg`: `cfg()` (instead of `values()` and `names()` being incomplete and subtlety incompatible with each other).
+
+## A look at the feature 
+
+Every time a Cargo feature is declared that feature is transformed into a config that is passed to `rustc` (the Rust compiler) so it can verify with its [well known cfgs](https://doc.rust-lang.org/nightly/rustc/check-cfg.html#well-known-names-and-values) if any of the `#[cfg]`, `#![cfg_attr]` and `cfg!` have unexpected configs and report a warning with the `unexpected_cfgs` lint.
+
+*`Cargo.toml`*:
+    
+```toml
+[package]
+name = "foo"
+
+[features]
+lasers = []
+zapping = []
+```
+
+*`src/lib.rs`:*
+
+```rust
+#[cfg(feature = "lasers")]  // This condition is expected
+                            // as "lasers" is an expected value
+                            // of the `feature` cfg
+fn shoot_lasers() {}
+
+#[cfg(feature = "monkeys")] // This condition is UNEXPECTED
+                            // as "monkeys" is NOT an expected
+                            // value of the `feature` cfg
+fn write_shakespeare() {}
+
+#[cfg(windosw)]             // This condition is UNEXPECTED
+                            // it's supposed to be `windows`
+fn win() {}
+```
+
+*`cargo check`*:
+
+![cargo-check](../../../../images/2024-05-05-check-cfg/cargo-check.svg)
+
+## Custom cfgs and build scripts
+
+> In Cargo point-of-view: a custom cfg is one that is neither defined by `rustc` nor by a Cargo feature. Think of `tokio_unstable`, `has_foo`, ... but not `feature = "lasers"`, `unix` or `debug_assertions`
+
+Some crates use custom cfgs that they either expected from the environment (`RUSTFLAGS`or other means) or is enabled by some logic in the crate `build.rs`. For those crates Cargo provides a new instruction: [`cargo::rustc-check-cfg`](https://doc.rust-lang.org/nightly/cargo/reference/build-scripts.html#rustc-check-cfg)[^2] (or `cargo:rustc-check-cfg` for older Cargo version).
```

Comment:
Unsure if we want to mention this. The double colons version of `cargo::rustc-check-cfg` should have been supported before rustc `--check-cfg` hits stable.

The only reason this is needed is someone gets stuck in a older version of nightly toolchain.

Reply:
> The only reason this is needed is someone gets stuck in a older version of nightly toolchain.

I think it is also useful for old stable, since users might actively support old stable versions and they might not want warnings if they can avoid it. Idk maybe I wrong.

I can remove it, if desired.

Reply:
True. Thanks for the clarification.